### ISSUE TITLE
CA1062 Validate Arguments of Public Methods False Positive

### ISSUE THREAD

**Issue by brennadev** [dotnet/roslyn-analyzers#3240]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

[CA1062](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1062?view=vs-2019)

#### Repro steps
In the code below, the call to this.Test in Test2 causes CA1062 to appear (line 3), but the parameter in Test is confirmed to not be null before anything is done with it, thus causing a false positive.

```
public void Test2(String value)
{
      this.Test(value);
}
```

```
private void Test(String value)
{
        var ok = value != null;
        if (ok)
        {
            Console.WriteLine(value.Length);
        }

        // do more with var ok
}
```
#### Expected behavior
No error due to CA1062 being violated

#### Actual behavior
CA1062 shows as being violated

**Comment by brennadev** [comment]
I came across another case where CA1062 appears to give a false positive, and it may be related to the scenario in the original description. CA1062 occurs on line 3, but in Bar, it's known that if `test` is not null, then `value` is not null on line 19 (the Console.WriteLine).

If this is separate from the original issue, I can move this into its own issue.

```
public void Foo(String value)
{
    this.Bar(value);
}

private void Bar(String value)
{
        String test = null;
        if (value != null)
        {
            test = "test";
        }

    // do something else

    if (test != null)
    {
        Console.WriteLine(value.Length);
     }
}
```

**Comment by Evangelink** [comment]
@mavasani I am not sure what could be done here except having the dataflow handling cross-procedural calls.

**Comment by mavasani** [comment]
@Evangelink we already perform interprocedural analysis.

**Comment by mavasani** [comment]
@brennadev this issue is very hard to tackle for any dataflow analysis as the nullness of two separate variables have been tied and you are relying on checking nullness of one to ensure non-null for other. I think it seems best for the code to perform explicit null check for value in the public method itself before passing it to another method.
